custom_components/ttlock/services.py

```diff
@@ -98,6 +111,26 @@ def _get_coordinators(self, call: ServiceCall) -> list[LockUpdateCoordinator]:
             ]
         return []
 
+    async def handle_list_passcodes(self, call: ServiceCall) -> ServiceResponse:
+        """List all passcodes for the given entities."""
+        passcodes = {}
+
+        for coordinator in self._get_coordinators(call):
+            codes = await coordinator.api.list_passcodes(coordinator.lock_id)
+            passcodes[coordinator.data.name] = [
+                {
+                    "name": code.name,
+                    "passcode": code.passcode,
+                    "type": code.type.name,
+                    "start_date": code.start_date.isoformat(),
+                    "end_date": code.end_date.isoformat(),
```

Comment:
Did you try letting home assistants json serialiser handle this?
```suggestion
                    "start_date": code.start_date,
                    "end_date": code.end_date,
```